scio-elasticsearch2/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -211,34 +226,49 @@ public void processElement(ProcessContext c) throws Exception {
       private final ClientSupplier clientSupplier;
       private final SerializableFunction<T, Iterable<ActionRequest<?>>> toActionRequests;
       private final ThrowingConsumer<BulkExecutionException> error;
+      private final int maxBulkRequestSize;
 
       public ElasticsearchWriter(String clusterName,
                                  InetSocketAddress[] servers,
+                                 int maxBulkRequestSize,
                                  SerializableFunction<T, Iterable<ActionRequest<?>>> toActionRequests,
                                  ThrowingConsumer<BulkExecutionException> error) {
+        this.maxBulkRequestSize = maxBulkRequestSize;
         this.clientSupplier = new ClientSupplier(clusterName, servers);
         this.toActionRequests = toActionRequests;
         this.error = error;
       }
+
       @ProcessElement
       public void processElement(ProcessContext c) throws Exception {
-        final List<Iterable<ActionRequest<?>>> actionRequests =
-            StreamSupport.stream(c.element().getValue().spliterator(), false)
-                .map(toActionRequests::apply)
-                .collect(Collectors.toList());
+        final Iterable<T> values = c.element().getValue();
 
         // Elasticsearch throws ActionRequestValidationException if bulk request is empty,
         // so do nothing if number of actions is zero.
-        if (actionRequests.isEmpty()) {
+        if (values.iterator().hasNext()) {
           LOG.info("ElasticsearchWriter: no requests to send");
           return;
         }
 
-        final BulkRequest bulkRequest = new BulkRequest().add(Iterables.concat(actionRequests));
-        final BulkResponse bulkItemResponse = clientSupplier.get().bulk(bulkRequest).get();
-        if (bulkItemResponse.hasFailures()) {
-          error.accept(new BulkExecutionException(bulkItemResponse));
-        }
+        final Stream<ActionRequest> actionRequests =
+            StreamSupport.stream(values.spliterator(), false)
+                .map(toActionRequests::apply)
+                .flatMap(ar -> StreamSupport.stream(ar.spliterator(), false));
+
+        final Iterable<List<ActionRequest>> chunks =
+            Iterables.partition(actionRequests::iterator, maxBulkRequestSize);
```

Comment:
@nevillelyh It's the underlying implementation for it https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/Iterables.java#L526